audit-service/src/main/java/uk/gov/gchq/palisade/service/audit/request/AuditMessage.java

```diff
@@ -0,0 +1,230 @@
+
+import uk.gov.gchq.palisade.Context;
+import uk.gov.gchq.palisade.Generated;
+import uk.gov.gchq.palisade.resource.LeafResource;
+import uk.gov.gchq.palisade.rule.Rules;
+import uk.gov.gchq.palisade.service.audit.request.common.domain.User;
+
+import java.util.Objects;
+import java.util.StringJoiner;
+
+/**
+ * Audit information for a request provided by each service.
+ * Each individual service sends a record to the Audit Service for every request that it receives.
+ * The components of the message will differ depending on which service has sent the data and if the processing was
+ * successful or not.
+ *
+ */
+@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+public class AuditMessage {
+
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+
+
+    public final String timeStamp; //when the service processed the request
+    public final String serverIp;   //server for the service that processed the request
+    public final String serverHostname;  //server host name for the service
+
+    public final Context context;  //Context of the client's request
+
+    public final String userId;    //User Id for the client.  Can be null if there is a User
+    public final User user;        //User  for the client.  Can be null if there is a userId
+
+    public final String resourceId;  //Resource Id for the client.  Can be null if there is a Resource
+    public final LeafResource resource; //Resource for the client.  Can be null if there is a Resource Id
+
+    public final Rules rules;   //Rules that apply to the request.  Can be null if this has not been generated yet.
+    public final long numberOfRecordsReturned; //Number of records that are provided for the request.  Can be 0 if this has not been generated yet.
+    public final long numberOfRecordsProcessed; //Number of records that have been processed for the request.  Can be 0 if this has not been generated yet.
+
+    public final String errorMessage;  //Error message that occurred during the processing of the request.  Will be null if there was no issue.
+
+    @SuppressWarnings("checkstyle:parameterNumber")
+    @JsonCreator
+    private AuditMessage(
+            final @JsonProperty("timeStamp") String timeStamp,
+            final @JsonProperty("serverIp") String serverIp,
+            final @JsonProperty("serverHostname") String serverHostname,
+            final @JsonProperty("context") Context context,
+            final @JsonProperty("userId") String userId,
+            final @JsonProperty("user") User user,
+            final @JsonProperty("resourceId") String resourceId,
+            final @JsonProperty("resource") LeafResource resource,
+            final @JsonProperty("rules") Rules rules,
+            final @JsonProperty("numberOfRecordsReturned") long numberOfRecordsReturned,
+            final @JsonProperty("numberOfRecordsProcessed") long numberOfRecordsProcessed,
+            final @JsonProperty("errorMessage") String errorMessage) {
```

Comment:
There should be several classes of AuditMessage, rather than cramming all the info into one do-all message... What would the palisade-service report as numberOfRecordsProcessed/Returned? What would the data-service report as the resourceId? What would a non-error reports as an errorMessage? Keep in 4 separate classes of data-request/data-read was successful/error

Reply:
There are several classes of AuditMessage, one for each service.  In each instance the class only knows about the objects that exists for that service.  The one at the audit-service takes whatever it given with some required and other optional.  One key to this exercise is to flatten and simplify objects.  The idea is to treat the message types as carriers of information.  Keep the logic in the services.  In this way there is no need for a set of classes and the related inheritance to process data.  That issue is deferred to the services.

Reply:
Complete overhaul of AuditMessage. Have another look